lib/common/socket/evloop.c.h

```diff
@@ -535,6 +535,37 @@ static void run_pending(h2o_evloop_t *loop)
     }
 }
 
+void h2o_evloop_destroy(h2o_evloop_t *loop)
+{
+    struct st_h2o_evloop_socket_t *sock;
+    //dispose all socket
+    while (loop->_pending_as_server != NULL || loop->_pending_as_client != NULL) {
+        while ((sock = loop->_pending_as_client) != NULL) {
+            loop->_pending_as_client = sock->_next_pending;
+            sock->_next_pending = sock;
+            do_dispose_socket((h2o_socket_t *)sock);
+        }
+        if ((sock = loop->_pending_as_server) != NULL) {
+            loop->_pending_as_server = sock->_next_pending;
+            sock->_next_pending = sock;        
+            do_dispose_socket((h2o_socket_t *)sock);
+        }
+    }
+      //looks like no need to free  _timeouts 
+      //h2o_timeout__do_dispose(loop,  loop->_timeouts);
```

Comment:
You are correct that the timeouts cannot be freed here. It is because `h2o_timeout_t` is owned by the application, not the event loop.

The right thing to do here would be: `assert(h2o_linklist_is_empty(&loop->_timeouts))` to check that application has released destroyed all the timers before calling `h2o_evloop_destroy`.